### PR TITLE
Feature/7601 clean up web job

### PR DESCRIPTION
**Pull request by ganpat-tech**
Code Changes
Unit test Cases

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+                            foreach (DirectoryInfo dir in di.GetDirectories())
+                            {
+                                var batchId = dir.Name;
+                                string scsResponseFileName = new DirectoryInfo(batchId).Name + ".json";
+                                var response = await azureBlobStorageClient.DeleteFileFromContainer(storageAccountConnectionString, containerName, scsResponseFileName);
+                                if (response)
+                                {
+                                    dir.Delete(true);
+                                    logger.LogInformation(EventIds.HistoricDateFolderDeleted.ToEventId(), "SCS response json file {ScsResponseFileName} deleted successfully from the container.", scsResponseFileName);
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs:56]: Kindly change eventid to HistoricSCSResponseFileDeleted

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageClient.cs [hunk 2]
```diff
@@ -32,5 +32,13 @@ public async Task<string> DownloadTextAsync(CloudBlockBlob cloudBlockBlob)
         {
              return await cloudBlockBlob.DownloadTextAsync();
         }
+        public async Task<bool> DeleteFileFromContainer(string storageAccountConnectionString, string containerName, string batchId)
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageClient.cs:35]: Remove this method as you can reuse existing method GetCloudBlockBlob of same class.
Get the blob and then delete.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+            var response = await exchangeSetCleanUpService.DeleteHistoricFoldersAndFiles();
+
+            Assert.AreEqual(false, response);
+        }
+
+        [Test]
+        public async Task WhenHistoricFoldersAndFilesFound_ThenReturnTrueResponse()
+        {
+            fakeConfiguration["HOME"] = @"D:\\Downloads";
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs:68]: Kindly use existing variable fakeFilePath
fakeConfiguration["HOME"] = fakeFilePath;

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Configuration/CleanUpConfig.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace UKHO.ExchangeSetService.CleanUpJob.Configuration
+{
+    public class CleanUpConfig
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Configuration/CleanUpConfig.cs]: As per suggestion, i have done required name changes.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                var subFolder = Directory.GetDirectories(filePath);
+                var historicDateTimeInUtc = DateTime.UtcNow.AddDays(-numberOfDays);
+                DateTime historicDate = new DateTime(historicDateTimeInUtc.Year, historicDateTimeInUtc.Month, historicDateTimeInUtc.Day);
+
+                foreach (var subFolderName in subFolder)
+                {
+                    DateTime subFolderDateTime;
+                    string dateFolder = new DirectoryInfo(subFolderName).Name;
+                    bool isValidDate = DateTimeExtensions.IsValidDate(dateFolder, out subFolderDateTime);
```
[the remark this answers is not in the document]
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs]: As per suggestion, i have done required name changes.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageClient.cs (hunk 2 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageClient.cs]: As per suggestion, i have done required changes and removed this method.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs (hunk 1 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs]: i have done changes eventid to HistoricSCSResponseFileDeleted.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs (hunk 3 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs]: As per suggestion, i have used existing variable fakeFilePath.

**Review by janvi13603 — APPROVED** [review]
Thanks for fixing review comments.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                return true;
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.DeleteHistoricFoldersAndFilesException.ToEventId(), ex, "Exception while deleteing historic folders and files with error {Message}", ex.Message);
+                return false;
+            }
+        }
+    }
+}
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs:88]: There appears to be quite a bit of cyclic complexity in this method. Would it be possible to break it down into a number of methods? Each method could be named to describe the work its is doing which might make it easier to follow the logic/flow.

Interested to know if this piece of code needs to be excluded from code coverage as it feels like it is doing quite a bit of work with a number of steps which could go wrong/could be tested?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        [Test]
+        public void WhenScsStorageAccountAccessKeyValueNotFound_ThenReturnKeyNotFoundException()
+        {
+            A.CallTo(() => fakeScsStorageService.GetStorageAccountConnectionString(A<string>.Ignored, A<string>.Ignored))
+              .Throws(new KeyNotFoundException("Storage account accesskey not found"));
+
+            Assert.ThrowsAsync(Is.TypeOf<KeyNotFoundException>()
+                   .And.Message.EqualTo("Storage account accesskey not found")
+                    , async delegate { await exchangeSetCleanUpService.DeleteHistoricFoldersAndFiles(); });
+        }
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs:52]: Is this test not testing the AzureFileSystemHelper - see comments made earlier on the AzureFileSystemhelper method!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            fakeConfiguration["HOME"] = fakeFilePath;
+
+            A.CallTo(() => fakeScsStorageService.GetStorageAccountConnectionString(A<string>.Ignored, A<string>.Ignored)).Returns(fakeStorageAccountConnectionString);
+            A.CallTo(() => fakeAzureFileSystemHelper.DeleteDirectoryAsync(fakeCleanUpConfig.Value.NumberOfDays, fakeStorageAccountConnectionString, fakeStorageConfig.Value.StorageContainerName, fakeFilePath)).Returns(true);
+
+            var response = await exchangeSetCleanUpService.DeleteHistoricFoldersAndFiles();
+
+            Assert.AreEqual(true, response);
+        }
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs:76]: I think the two tests above re actually testing that the call azureDeleteFileSystemHelper.DeleteDirectoryAsync takes place (which is fine).

However the result of the call should ideally be tested on the azureDeleteFileSystemHelper (see earlier comments)

**Review by martyn-fewtrell — COMMENTED** [review]
On the whole a good PR.

A couple of questions about the logic and tests on the ExchangeSetCleanUpService & AzureFileSystemHelper which could be my understanding or may require tweaking.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs (hunk 8 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs:88]: Resolved, I have made required changes for same.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs (hunk 7 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.CleanUpJob.UnitTests/Services/ExchangeSetCleanUpServiceTest.cs:57]: Resolved, As discussed we have added FakeAzureFileHelper for checking methods which actually calls and returns bool value from that.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs (hunk 6 above)
Comment (ganpat-tech) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.CleanUpJob/Helpers/AzureFileSystemHelper.cs]: Resolved. As discussed we have break it down into a number of methods.
Yes for this piece of code its need physical file path when we are doing unit testing so for this reason we have excluded from code coverage.

**Review by martyn-fewtrell — APPROVED** [review]
Thanks for the updates

**Pull request merged**